design-proposals/klip-2-produce-data.md

```diff
@@ -0,0 +1,347 @@
+After installing KSQL, developers are met with an immediate hurdle - they must have existing data 
+in Kafka that match the data they expect to handle. In a relational database environment, developers
+can use `INSERT INTO ... VALUES` to quickly construct a toy environment for testing. Generating this 
+data KSQL, however, is cumbersome: 
+
+* you require a separate tool (e.g. `kafkacat`, `ksql-datagen`)
+* you must know the deployment topology (e.g. URIs for ZK/Broker)
+* there is no synergy with KSQL (e.g. tools are not *schema aware* and can produce arbitrary data)
+
+This KLIP attempts to integrate producing test data into the KSQL REPL.
+
+## Scope
+
+This KLIP is intended to improve the *development and testing* workflow as opposed to providing
+*production-ready* functionality. The following tasks will be addressed:
+
+* KSQL will be able to create and produce data to arbitrary topics
+* KSQL will be able to validate data when producing to a topic that is registered with a schema
+in KSQL (e.g. as a result of `CREATE STREAM`)
+
+In the future, this functionality can be extended to be "production grade", allowing system
+operators or first-responders to quickly patch up data holes in case of emergency. This KLIP will
+explicitly discourage such usage in documentation (as it needs further testing before we can
+recommend it), but will not artificially impose a restriction on such usage.
+
+## Value
+
+This proposal will improve the first 15-minute experience that developers have with KSQL and will
+help seasoned KSQL developers test new functionality. Furthermore, this functionality can be
+leveraged to create a powerful testing harness for KSQL applications that can run in a sandbox
+without pre-existing data.
+
+## Public APIS
+
+#### INSERT INTO ... VALUES
+
+```sql
+INSERT INTO <stream_name|table_name> [(column1, column2, ...)]
+  VALUES (value1, value2, ...)
+```
+
+The `INSERT VALUES` statement can be used to insert new records into a KSQL source (agnostic to the
+Stream/Table duality). If the columns are not present, it is assumed that the values are in the same
+order as the schema and contain every field. `ROWKEY` is always present as the first column, and is 
+automatically populated with semantics described in the section on `ROWKEY` and `ROWTIME` below.
+
+The value for `stream_name`/`table_name` must be a valid Stream/Table registered with a KSQL schema.
+The serialization format will be the same as the format specified in the `value_format` of the
+`CREATE` statement. 
+
+For the initial version, only primitive literal types will be supported (e.g. `INT`, `VARCHAR`, 
+etc...) but syntax is easily extensible to support struct, map and list (see _Complex Datatypes_
+section below).
+
+Future extensions: 
+* Support Structs/Complex Types
+* Support some `AT INTERVAL` syntax to generalize the functionality to generate full stream data.
+* Read from file to allow reproducible insertions
+* Headless mode support
+
+#### DELETE FROM
```

Comment:
Good point - I didn't think this one through. We could support `DELETE` on a stream to solve that, but that seems a little weird to me since streams are basically append-only logs.

Reply:
I just verified with @hjafarpour - looks like we completely ignore `null` values in streams today (doesn't make sense from a SQL perspective) so I think we're OK leaving this unsupported

Reply:
Humm.... you can treat a tables commit log as a stream or materialize a stream into a table, (i.e stream-table duality). In both cases the stream can contain records with a null value: the table commit log (which is a stream) can contain null values to indicate a delete and for the materialized stream the nulls result in values from the table being removed.

I _guess_ we could say that if you want to introduce a tombstone into a stream then you _must_ import it as a table, i.e. if you've already imported it as a stream you'll have to re-import it as a table.  But I'm not sure what we gain by the restriction.

Reply:
@hjafarpour be interested in your opinion on the above...

Reply:
I think after talking to @blueedgenick and @MichaelDrogalis I've decided to just punt the DELETE conversation to another time and will not implement it in V1.